### PR TITLE
dronegen: Add tag builds

### PR DESCRIPTION
**Pull request by webvictim**
Implements tag builds using dronegen. Mostly just looking for style/flow comments.

`.drone.yml` diff has been validated using https://github.com/gravitational/teleport/commit/4beafe5bfdec77d107c28e4c04aaca8626a51c92
- Also renames `build.go` to `tag.go` to be a bit more descriptive.
- d5d9091 also contains the fixes to remove the old `test-docs-internal`/`test-docs-external` pipelines and implement the `test-docs` pipeline instead.

### CODE REVIEW

File: dronegen/push.go
Comment (a-palchikov) [inline dronegen/push.go:39]: nit: any reason this won't use `make print-version` to extract the version?

File: dronegen/tag.go
Comment (a-palchikov) [inline dronegen/tag.go:312]: nit: since this comes up on multiple occasions, this could potentially be factored out as `volumeDocker`

File: dronegen/push.go
Comment (webvictim) [reply dronegen/push.go:39]: Embarrassingly, I hadn't even noticed that was a thing. I think originally I was doing everything to avoid installing `make` inside every single Docker image, but that does seem quite a lot simpler to use.

**Comment by webvictim** [comment]
@awly I've validated that all the YAML diffs look good side-by-side - would appreciate a review on the Go style if you could please. Once this PR looks good to you, I can merge it into `andrew/dronegen` and then fix that whole PR up and get it ready to merge by rebasing against `master` if you like.

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:32]: can this be combined with `pushBuildType`, if `push.go` just never sets `params.centos6` to `true`?

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:68]: curious - why is this only for FIPS builds?

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:110]: can renaming be handled inside the makefile?

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:178]: can this title be removed?
pipeline name should be clear enough on its own, steps can be just "build" and "copy"

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:202]: this should be a constant in `common.go`, re-used by push pipelines too
ideally, loaded at runtime from `$RUNTIME` var in `build.assets/Makefile` if there's a magic command to print it out

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:216]: should this be in `common.go` and used by other pipelines?

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:282]:
```suggestion
	artifactOSS := true
	artifactType := fmt.Sprintf("%s-%s", params.os, params.arch)
	if params.fips {
		artifactType += "-fips"
		artifactOSS = false
	}

	if artifactOSS {
		commands = append(commands,
			fmt.Sprintf(`aws s3 cp s3://$AWS_S3_BUCKET/teleport/$${S3_PATH}teleport-v$${VERSION}-%s-bin.tar.gz /go/artifacts/`, artifactType),
		)
	}
	commands = append(commands,
		fmt.Sprintf(`aws s3 cp s3://$AWS_S3_BUCKET/teleport/$${S3_PATH}teleport-ent-v$${VERSION}-%s-bin.tar.gz /go/artifacts/`, artifactType),
	)
```

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:339]: make it a `switch` and panic in `default` case

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:375]: should this be just `volumeRefDocker`?

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:422]: optional: extract this into `func uploadToS3Step(s s3Settings) step`, using
```
type s3Settings struct {
  region string
  source string
  target string
  stripPrefix string
}
```
and inject the secret values automatically

File: dronegen/tag.go
Comment (webvictim) [reply dronegen/tag.go:68]: It's a byproduct of the fact that `e/Makefile` doesn't set `VERSION` itself; it relies on it being inherited from the OSS `Makefile` in the root directory because `e` is intended to be a submodule and doesn't have a way to set `VERSION` as a standalone checkout. As FIPS builds are Enterprise-only the variable doesn't get inherited.

We could actually set `VERSION` using this method for every single build with no negative consequences, but I'm trying to avoid introducing additional changes as part of the `dronegen` migration. We can look to make improvements later.

File: dronegen/tag.go
Comment (webvictim) [reply dronegen/tag.go:110]: Theoretically yes, we could make this change there. As I mentioned above I'm trying to avoid making changes to anything outside of Drone via this PR as much as possible - I'll look to fix these more cosmetic things later as part of a follow-up PR to avoid complications.

File: dronegen/tag.go
Comment (webvictim) [reply dronegen/tag.go:178]: Yeah, this is basically a hangover from when I was validating the YAML by eye, I used it to make sure that I'd actually copied the correct block for RPM vs DEB vs tarball etc.

File: dronegen/tag.go
Comment (webvictim) [reply dronegen/tag.go:202]: I'll move it to `common.go` and reuse it, but for now we'll leave it statically defined so we don't introduce too many changes at once.

File: dronegen/tag.go
Comment (webvictim) [reply dronegen/tag.go:216]: I think I was avoiding this so far because `services` and `volumes` are tightly coupled which means there's a minimum of two common structs and it didn't seem entirely worth it.

There are also three slightly different versions of this used:
- one for regular `docker:dind` with `/var/run` (most common)
- one with the above plus `/tmpfs` for in-memory operations like lint/unit/integration tests
- one with both of the above plus `/tmp` for a second `emptyDir` during integration tests

I'll see if I can make the common use case simpler.

File: dronegen/tag.go
Comment (webvictim) [reply dronegen/tag.go:375]: When building RPMs you need access `tmpfs` for the key material as well, so it's special-cased.

**Comment by webvictim** [comment]
@awly PTAL

File: dronegen/common.go
Comment (awly) [inline dronegen/common.go:45]: Change this to 
```go
func dockerService(extraVolumes ...volumeRef) service {
	return service{
		Name:    "Start Docker",
		Image:   "docker:dind",
		Volumes: append(extraVolumes, volumeRefDockersock),
	}
}
```
to make this a bit more flexible.
Then you can drop the `extraVolumes` struct and call this as:
`docerService(volumeRefTmpfs, volumeRefDind)`

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:24]: Rename this to `releaseMakefileTarget` and remove `pushMakefileTarget`

File: dronegen/tag.go
Comment (awly) [inline dronegen/tag.go:174]:
```suggestion
	for _, arch := range []string{"amd64", "386", "arm", "arm64"} {
		for _, fips := range []bool{false, true} {
			if (arch == "386" || arch == "arm") && fips {
				// FIPS mode not supported on i386 or ARM
				continue
			}

			ps = append(ps, tagPipeline(buildType{os: "linux", arch: arch, fips: fips}))

			if arch == "arm" || arch == "arm64" {
				// TODO(gus): support needs to be added upstream for building ARM/ARM64 packages first
				continue
			}
			for _, packageType := []string{rpmPackage, debPackage} {
				ps = append(ps, tagPackagePipeline(packageType, buildType{os: "linux", arch: arch, fips: fips}))
			}
		}
	}
```

**Comment by webvictim** [comment]
@awly PTAL!

**Review by awly — APPROVED** [review]

**Comment by webvictim** [comment]
Closing in favour of #6071

**Pull request closed without merging**
